rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,276 @@
+---
+authors: Alexander Klizhentas (sasha@goteleport.com), Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 59 - Search Based Access Requests
+
+## What
+
+This RFD proposes just-in-time search-based access requests which will enable
+users to request access to resources based on a search and/or selection of
+individual resources, rather than having to request one or more roles.
+
+## Why
+
+Teleport users may not know in advance which roles they need or who should grant
+access, just which resource(s) they need access to.
+
+Users would like to avoid having many different custom roles for different
+levels of access.
+
+Requesting access to a "blanket" role violates the principle of least priviledge.
+Most of the time incident responders only need access to 1 or 2 nodes,
+search-based access requests will allow them to request access to only the nodes
+they need.
+
+
+## Details
+
+Users will be able to request access by using search, and roles will be used
```

Comment:
I think this is what you are proposing below, but its worth calling it out here: our preference is that users are not just able to request access by search - they are able to request access by simply connecting  to a host. The tsh infra then does the search for them and decides if they are in this new category of "can access, but with permission from <x slack room>". 

If they are in that group, they are then told that access to that resource requires a reason and permission, and the tsh ssh command just blocks waiting for that to be approved or denied. 

The user experience (i.e. tsh ssh <host> blocking) is much more important than how we wire up deciding who is in that group of 'sure, but <x> slack room needs to approve', since the former affects the usability for users (i.e. whether we can actually use this) whereas the latter is simply a question of what API we have to deal with (we can deal with any reasonable API).

Reply:
The original idea was for access requests based on search and that lead to the way this document was originally written, but automatically requesting access just by attempting to connect to a host is very powerful and may well end up being the more popular usecase. I will make sure it's a first-class feature here particular attention on the UX.